### DISCUSSION TITLE
Integration with EMF.cloud Model Server on Eclipse IDE

### DISCUSSION THREAD

**Discussion by Avaritia1** [eclipse-glsp/glsp#994] [Q&A]
Hi! Currently, I've created a GLSP server using the [template](https://github.com/eclipse-glsp/glsp-examples/tree/master/project-templates/java-emf-eclipse) running on Eclipse based on Java and EMF. I now want to implement a GLSP server based on Java and EMF.cloud running on Eclipse. The idea is to build a graphical editor based on my model with connection to the server. But I found that there is no such combination template. I wonder if such a combination is possible? Or can you give me some suggestions or ideas to realize this integration? Thanks!

**Comment by ndoschek** [comment]
Hi @Avaritia1,

Thanks for your interest :tada:! Yes, there is one project-template available at https://github.com/eclipse-emfcloud/modelserver-glsp-integration/tree/main/project-templates/modelserver-glspjava-emf-theia.
Please note: Currently, there is one PR open, which updates to the latest GLSP RC version (see https://github.com/eclipse-emfcloud/modelserver-glsp-integration/pull/56).

HTH and best regards,
Nina

**Reply by Avaritia1** [reply]
Thanks for the reply! @ndoschek But I have tried this template before, I understand this template is based on Eclipse Theia. I wonder if it can work with Desktop Eclipse. The idea is that I want to build a web-based graphical editor that enables interaction between two people using Desktop Eclipse on two computers. If it is doable, can you give me some suggestions or ideas? Thanks!

**Reply by ndoschek** [reply]
Hi @Avaritia1, oh sorry I overlooked that earlier!
Overall, you could try to combine both project-templates, and use use the backend part with both server components from https://github.com/eclipse-emfcloud/modelserver-glsp-integration/tree/main/project-templates/modelserver-glspjava-emf-theia/server and the frontend integration part from https://github.com/eclipse-glsp/glsp-examples/tree/master/project-templates/java-emf-eclipse/glsp-client.

Until now, we didn't come across the multi user use case with the Eclipse integration.
Is there a specific reason why you would like to stick to the Eclipse IDE as platform?

In the meantime, I can point you to another project of ours, based on Theia, which perhaps might be of interest to you:
The [coffee-editor](https://www.eclipse.org/emfcloud/#coffeeeditoroverview) is a comprehensive example of a web-based modeling tool based on EMF.cloud, including a form-based editor based on JSON Forms, a diagram based on GLSP, a textual DSL, a code generator and a graphical visualization of an analysis result using D3. The coffee editor uses the model sever for synchronizing all editors and it is embedded into the Eclipse Theia IDE.
You can try it out here:  https://try.theia-cloud.io/?appdef=coffee-editor

Best, Nina

**Reply by Avaritia1** [reply]
Thanks for the suggestions! @ndoschek I will try the idea to combine two project-templates. 

The thing is our project is based on Eclipse IDE. Propagation of semantic changes such as adding, deleting, editing model text is currently supported. Therefore, my goal is to build a graph editor module that accommodates non-semantic changes, such as: movement of nodes, generation of edges, etc. I think GLSP provides a good template and I want to explore its integration with Eclipse IDE. I found that this  kind of integration currently only supports EMF, but my project needs the content of EMF.cloud. So I plan to explore this new combination a little bit :)

**Reply by ndoschek** [reply]
Okay that sounds interesting! Hope it works out!
In case there arise any questions, just let us know :)

Best, Nina

**Reply by Avaritia1** [reply]
Hi! @ndoschek Now I am deploying the [emf.cloud project template](https://github.com/eclipse-emfcloud/modelserver-glsp-integration/tree/main/project-templates/modelserver-glspjava-emf-theia) on Theia. However I found that there was an error when I run `yarn build`. This error didn't happen before. I remembered there was a "issues 55" branch three weeks ago. And I noticed there was a commit last week, so I wonder if there is still a problem in the existing project template. Thanks in advance!
<img width="1470" alt="e75ca1178cbb9fd7991d217859f2858" src="https://github.com/eclipse-glsp/glsp/assets/80332537/22f3ae2d-5427-4036-b39c-5bfdc92ee29e">

**Reply by Avaritia1** [reply]
Small addition: It is strange that it works on my Windows PC :) However, after I run `yarn build` and `yarn start`, and launch the example in the browser on (http://localhost:3000/). The example.tasklist doesn't show anything. There's just an exclamation point sign saying "initializing". I guess there should be a graphical editor similar to the Eclipse project, right?
![image](https://github.com/eclipse-glsp/glsp/assets/80332537/d651f4ff-7687-4d3f-baf9-a9aeb5134760)
![image](https://github.com/eclipse-glsp/glsp/assets/80332537/9a334436-3964-45c6-b440-59a490dfd79f)

**Reply by ndoschek** [reply]
Hi @Avaritia1,
The last commit updated the underlying frameworks to the latest RC versions (e.g. GLSP and the Model Server dependencies).
With a clean state there should not be any build errors, considering the project's [prerequisites listed in the README](https://github.com/eclipse-emfcloud/modelserver-glsp-integration/tree/main/project-templates/modelserver-glspjava-emf-theia#prerequisites).

Could you try to do a fresh clone of the repo or do a `git clean -xdf` and see if the build works then for you?

Best, Nina

**Reply by ndoschek** [reply]
Oh, i just saw your additional reply. Did the build work for you now?

Yes, the [modelserver-glsp-integration project](https://github.com/eclipse-emfcloud/modelserver-glsp-integration/tree/main/project-templates/modelserver-glspjava-emf-theia) will show the same diagram.

**Reply by Avaritia1** [reply]
Thanks! What I did was downloading the zip file of the latest repo instead of using git. Is that the reason that cause the problem? And I tried again on my Windows PC, it still showed "initializing" instead of the diagram.

**Reply by Avaritia1** [reply]
And when I run `yarn start` and open (http://localhost:3000/), I find those errors in the terminal. Are these errors related to browser version?
![image](https://github.com/eclipse-glsp/glsp/assets/80332537/de331feb-6217-4a3f-bb5d-96985734ed5c)

**Reply by ndoschek** [reply]
Hi again @Avaritia1,
Thanks for the additional information!
I could spot the issue and already pushed a fix to the repository, if you update your repository it should work again!
I usually test with the server products in backend, hence I didn't come across this issue unfortunately (which only occurred if the servers are started as JAR products).
But now it should work again either way :)
Best, Nina

**Reply by Avaritia1** [reply]
Thanks for the help! It does solve the "initializing" problem. However, it shows another problem this time:
`Error during model loading
org.eclipse.glsp.server.types.GLSPServerException: Error during model loading
	at org.eclipse.emfcloud.modelserver.glsp.EMSModelServerAccess.getSemanticModel(EMSModelServerAccess.java:82)
	at org.eclipse.emfcloud.modelserver.glsp.notation.integration.EMSNotationSourceModelStorage.doLoadSourceModel(EMSNotationSourceModelStorage.java:29)
	at org.eclipse.emfcloud.modelserver.glsp.EMSSourceModelStorage.loadSourceModel(EMSSourceModelStorage.java:33)
	at org.eclipse.glsp.server.features.core.model.RequestModelActionHandler.executeAction(RequestModelActionHandler.java:64)
	at org.eclipse.emfcloud.modelserver.glsp.actions.handlers.EMSRequestModelActionHandler.executeAction(EMSRequestModelActionHandler.java:31)
	at org.eclipse.emfcloud.modelserver.glsp.actions.handlers.EMSRequestModelActionHandler.executeAction(EMSRequestModelActionHandler.java:22)
	at org.eclipse.glsp.server.actions.AbstractActionHandler.execute(AbstractActionHandler.java:53)
	at org.eclipse.glsp.server.internal.actions.DefaultActionDispatcher.runAction(DefaultActionDispatcher.java:193)
	at org.eclipse.glsp.server.internal.actions.DefaultActionDispatcher.handleAction(DefaultActionDispatcher.java:172)
	at org.eclipse.glsp.server.internal.actions.DefaultActionDispatcher.handleNextAction(DefaultActionDispatcher.java:159)
	at org.eclipse.glsp.server.internal.actions.DefaultActionDispatcher.runThread(DefaultActionDispatcher.java:145)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.util.concurrent.ExecutionException: java.lang.RuntimeException: Could not parse 'data' field
	at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:395)
	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:1999)
	at org.eclipse.emfcloud.modelserver.glsp.EMSModelServerAccess.getSemanticModel(EMSModelServerAccess.java:79)
	... 11 more
Caused by: java.lang.RuntimeException: Could not parse 'data' field
	at org.eclipse.emfcloud.modelserver.internal.client.ModelServerClientDelegate.lambda$require$25(ModelServerClientDelegate.java:856)
	at java.base/java.util.Optional.orElseThrow(Optional.java:408)
	at org.eclipse.emfcloud.modelserver.internal.client.ModelServerClientDelegate.require(ModelServerClientDelegate.java:856)
	at java.base/java.util.function.Function.lambda$andThen$1(Function.java:88)
	at org.eclipse.emfcloud.modelserver.client.Response.body(Response.java:41)
	at org.eclipse.emfcloud.modelserver.internal.client.ModelServerClientDelegate.lambda$cacheModelType$3(ModelServerClientDelegate.java:205)
	at java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:642)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506)
	at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2073)
	at org.eclipse.emfcloud.modelserver.internal.client.ModelServerClientDelegate$3.onResponse(ModelServerClientDelegate.java:818)
	at okhttp3.internal.connection.RealCall$AsyncCall.run(RealCall.kt:519)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	... 1 more`
	
![image](https://github.com/eclipse-glsp/glsp/assets/80332537/227c39df-62f6-4d5a-ad53-b3e4a41a2c84)

**Reply by Avaritia1** [reply]
Hi! @ndoschek I wonder if you can help me with the error? Much appreciate!

**Reply by ndoschek** [reply]
Hi @Avaritia1!
I have a few questions:
- Did you re-build all components of the project-template either via `yarn build` or built the components separately?
- Do you use the provided workspace of the project-template? Do you have any changes on the model?
This error indicates that there is an error while loading the semantic model (`*.tasklist`) file.
Could you doublecheck if there aren't any changes on this file and try again?

I tested with latest master and the example should work as expected.
HTH and best regards, Nina

**Reply by Avaritia1** [reply]
Hi @ndoschek ! Thanks for the reply.
For question 1: Yes, I have re-build all components of the template. I tried both: `yarn build` and `yarn build:client` with  `yarn build:servers`. Both methods had the same error.
For question 2: I haven't changed the model. What I have done is downloading the zip file from Github and building it locally. And I checked the tasklist file, it is exactly the same as in Github.

**Reply by Avaritia1** [reply]
Small addition: I tried to build the template on my Mac, it worked :) So maybe it's a different system problem?

**Reply by ndoschek** [reply]
Glad to hear @Avaritia1!
Okay, perhaps it is really a Windows issue then, we use mainly Ubuntu on our side.
I can have a look next week on Windows and see if I can spot any inconsistencies.

However, I'm glad you got it to run! :)
Best, Nina

**Reply by Avaritia1** [reply]
Hi @ndoschek ! Thanks for the help. After my research, I feel that it is too cumbersome to realize the combination of the front end of Eclipse IDE and the back end of Theia project (many API problems, etc.). So I decided to continue my project on this [Theia based template](https://github.com/eclipse-emfcloud/modelserver-glsp-integration/tree/main/project-templates/modelserver-glspjava-emf-theia). 

The purpose of my project is to achieve collaboration, for example: first open an instance with localhost:3000, and then open an instance with localhost:3001. Both instances are based on the same model and use the model server for interaction. (for example, node movement on one instance will be received on the other instance). Do you think this idea is feasible? If so, can you provide some suggestions or ideas? Thanks!

**Reply by Avaritia1** [reply]
Hi @ndoschek! Do you have any suggestions or insights for my above thoughts? Any help would be much appreciated. Thanks in advance!

**Reply by ndoschek** [reply]
Hi @Avaritia1,
Glad to hear you continue your project with the EMF cloud project template!

In general, we didn't target the use case of supporting collaborative multi-user editing yet with the [EMF cloud Model Server project](https://github.com/eclipse-emfcloud/emfcloud-modelserver), yet.
It's main purpose is to support editing and synchronizing across different editing capabilities inside one IDE, e.g. diagram, form-based and property editing for example.

However, to achieve support for multiple users, I assume at least the subscription handling needs to be extended to deal with multiple users simultaneously. Because in the end to work on the same resources, it is necessary to work with the same unique identifiers (modelURIs) to broadcast all updates.
Currently, we have not planned to extend the project in that area, but also, we would gladly appreciate contributions :)

Best, Nina

**Reply by Avaritia1** [reply]
Hi @ndoschek, I have another question about this [template](https://github.com/eclipse-emfcloud/modelserver-glsp-integration/tree/main/project-templates/modelserver-glspjava-emf-theia).

For now, I am trying to use my customized model instead of tasklist model. I think there should be a place for source model file (like model.ecore), right? If not, how does this project load or process the source model?

Thanks in advance!

**Reply by ndoschek** [reply]
Hi @Avaritia1,
Yes that assumption is correct.
This template consumes its model from the same repository, [please see here](https://github.com/eclipse-emfcloud/modelserver-glsp-integration/tree/main/examples/org.eclipse.emfcloud.modelserver.glsp.example.model).
This is mainly, to be able to reuse it in multiple projects.

So if you want to use your own custom model, you can use that as guideline. You will have to create your own custom ecore model and genmodel and the resulting generated source code.
You can include this "model project" to your application project next to the server components for example.
In the next step you will then have to take care of proper resource management for your custom model in the Model Server.
For this you will need to adapt the `TaskListPackageConfiguration`, `TaskListResource`, `TaskListResourceSetFactory` (you may want to adapt the names according to your model). ThePackageConfiguration links the file extension to your custom model EPackage.
The ModelResourceManager should then be able to handle your resources accordingly.

The next step is then to adapt the GLSP server to be able to process the new source model and generate the graphical model.

Best, Nina
